### PR TITLE
sample: add numerical stability to softmax transform

### PR DESCRIPTION
**Pull request by ParthSareen**
- Possible that not applying numerical stability on softmax was creating issues on sampling
- collapsed the temperature and softmax transformations as both needed the max logit value

### CODE REVIEW

File: sample/samplers.go [hunk 1]
```diff
@@ -90,8 +90,7 @@ func (s *Sampler) sample(tokens []token) (token, error) {
 		sortLogits(tokens)
 	}
 
-	tokens = temperature(tokens, s.temperature)
-	tokens = softmax(tokens)
+	tokens = temperatureAndSoftmax(tokens, s.temperature)
```
Comment (jmorganca) [inline sample/samplers.go:93]: Do we want to keep calling this `temperature`? Is there a case where we we wouldn't want softmax?

Comment (ParthSareen) [reply sample/samplers.go:93]: should be okay to call temperature as that's the only param needed. need softmax in every case as the probabilities are needed for all subsequent steps

Comment (jmorganca) [reply sample/samplers.go:93]: I think we can name it `temperature` then for simplicity

File: sample/transforms.go [hunk 2]
```diff
@@ -5,41 +5,32 @@ import (
 	"slices"
 )
 
-func softmax(ts []token) []token {
+// apply temperature scaling and softmax to the logits
```
Comment (jmorganca) [inline sample/transforms.go:8]: GoDoc this

```
// temperature applies sacling and softmax to the logits
```

File: sample/samplers.go [hunk 3]
```diff
@@ -90,8 +90,9 @@ func (s *Sampler) sample(tokens []token) (token, error) {
 		sortLogits(tokens)
 	}
 
+	// token logit values are updated to probabilities
```
Comment (jmorganca) [inline sample/samplers.go:93]: not sure how helpful this comment would be

Comment (ParthSareen) [reply sample/samplers.go:93]: useful to know as topp and minp need probs

**Review by jmorganca — APPROVED** [review]
Some small things to address otherwise LGTM

File: sample/transforms_test.go [hunk 4]
```diff
@@ -53,11 +45,14 @@ func TestSoftmax(t *testing.T) {
...
 	}
 
-	// Check relative ordering is preserved
-	for i := 1; i < len(got); i++ {
-		if got[i].value < got[i-1].value {
-			t.Errorf("probability ordering not preserved at index %d", i)
-		}
+	got = temperature(toTokens(input), 1)
+	// Check probabilities sum to 1
+	sum = 0
```
Comment (jmorganca) [inline sample/transforms_test.go:50]:
```suggestion
	sum = 0.0
```
